# Data Exploration: Transactions

In [1]:
import pandas as pd
from helpers import RAW_DATA_DIR

## Exploration

In [2]:
df_transactions = pd.read_parquet(RAW_DATA_DIR / 'transactions_KL.parquet')
df_transactions

project_name    spa_date                   address  \
0         BANDAR BARU SRI PETALING  09/06/2023        ✕✕✕, JALAN PIKRAMA   
1         BANDAR BARU SRI PETALING  01/06/2023    ✕✕. ✕✕, JALAN PERLAK 3   
2         BANDAR BARU SRI PETALING  29/05/2023       ✕✕ ✕, JALAN 12/149L   
3         BANDAR BARU SRI PETALING  25/05/2023      ✕✕. ✕✕✕, JALAN PASAI   
4         BANDAR BARU SRI PETALING  22/05/2023  ✕✕, JALAN SRI PETALING 5   
...                            ...         ...                       ...   
294562      HERITAGE STATION HOTEL  13/11/1990   ✕✕✕-✕✕✕, BB WANGSA MAJU   
294563               IDAMAN PUTERI  10/01/2005        ✕✕-✕, JALAN GOMBAK   
294564         KELAB LE CHATEAU II  25/02/2008     ✕-✕✕-✕, JALAN KIARA 3   
294565  MUTIARA SENTUL CONDOMINIUM  10/08/2009   ✕-✕-✕, OFF JALAN SENTUL   
294566          TAI CHEONG MANSION  18/08/1995        ✕✕-✕✕-✕✕-✕, SENTUL   

                       building_type     tenure floors rooms  land_area  \
0       TERRACE HOUSE - INTERMEDIATE  LEASEHOLD      1   nan  2,196 ft²   
1       TERRACE HOUSE - INTERMEDIATE  LEASEHOLD      2   nan    753 ft²   
2       TERRACE HOUSE - INTERMEDIATE  LEASEHOLD     2½   nan  3,197 ft²   
3       TERRACE HOUSE - INTERMEDIATE  LEASEHOLD      2   nan    753 ft²   
4                             SEMI-D  LEASEHOLD     2½   nan  4,801 ft²   
...                              ...        ...    ...   ...        ...   
294562                          FLAT  LEASEHOLD      1     2    493 ft²   
294563                   CONDOMINIUM   FREEHOLD      1     3   1454 ft²   
294564                   CONDOMINIUM   FREEHOLD      1     3    593 ft²   
294565                     APARTMENT  LEASEHOLD      1     2   1193 ft²   
294566                          FLAT  LEASEHOLD      1     3   1927 ft²   

        built_up price_psf      price  
0            nan       342    750,000  
1            nan       398    300,000  
2            nan       188    600,000  
3            nan       531    400,000  
4            nan       250  1,200,000  
...          ...       ...        ...  
294562   493 ft²        71     35,000  
294563  1454 ft²       150    218,025  
294564   593 ft²       194    115,000  
294565  1193 ft²       197    235,000  
294566  1927 ft²       269    518,296  

[294567 rows x 11 columns]

In [3]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294567 entries, 0 to 294566
Data columns (total 11 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   project_name   294567 non-null  object
 1   spa_date       294567 non-null  object
 2   address        294546 non-null  object
 3   building_type  294567 non-null  object
 4   tenure         294567 non-null  object
 5   floors         294567 non-null  object
 6   rooms          294567 non-null  object
 7   land_area      294567 non-null  object
 8   built_up       294567 non-null  object
 9   price_psf      294567 non-null  object
 10  price          294567 non-null  object
dtypes: object(11)
memory usage: 24.7+ MB


At a glance:
1. `project_name` column contains strings. No missing values.
2. `spa_date` is detected as strings instead of date, need to change to date type. No missing values. 
3. `address` column contain strings. There are missing values.
4. `building_type` column contains strings. No missing values.
5. `tenure` column contains strings. No missing values.
6. `floors` is detected as strings instead of float, need to change to float type. No missing values.
7. `rooms` is automatically detected as float. There are 'nan' but not detected as missing values.
8. `land_area` is detected as strings instead of float, need to change to float type. No missing values.
9. `built_up` is detected as strings instead of float, need to change to float type. There are missing values.
10. `price` is detected as strings instead of float, need to change to float type. No missing values.

In [4]:
df_transactions['spa_date'] = pd.to_datetime(df_transactions['spa_date'], format='%d/%m/%Y')
df_transactions['spa_date']

0        2023-06-09
1        2023-06-01
2        2023-05-29
3        2023-05-25
4        2023-05-22
            ...    
294562   1990-11-13
294563   2005-01-10
294564   2008-02-25
294565   2009-08-10
294566   1995-08-18
Name: spa_date, Length: 294567, dtype: datetime64[ns]

In [5]:
df_transactions['spa_date'].sort_values()

56134    1909-05-11
128655   1933-10-25
262130   1955-07-25
64404    1959-07-27
266269   1960-11-14
            ...    
224137   2023-06-09
266662   2023-06-13
87238    2023-06-13
214691   2023-06-14
125442   2023-06-19
Name: spa_date, Length: 294567, dtype: datetime64[ns]

The data type of the `spa_date` column has been changed to datetime type.

In [6]:
print(f"% missing values: {df_transactions['address'].isnull().sum() / len(df_transactions) * 100}")
df_transactions[df_transactions['address'].isnull()]

% missing values: 0.0071291081485706145


project_name   spa_date address  \
18211                     ALAM DAMAI 1999-02-15    None   
27824             TAMAN SRI HARTAMAS 1996-01-10    None   
70435              KAMPUNG BATU MUDA 2000-06-06    None   
71295   LUCKY GARDEN (JALAN PUCHONG) 1997-04-21    None   
71296   LUCKY GARDEN (JALAN PUCHONG) 1997-04-21    None   
74962            SRI BINTANG HEIGHTS 1998-04-29    None   
75638                LAMAN RESIDENCE 2008-09-15    None   
102998             HAPPY GARDEN FLAT 1999-09-07    None   
112720            TAMAN BUKIT CHERAS 1998-03-20    None   
147816            SALAK SOUTH GARDEN 2002-10-29    None   
147843            SALAK SOUTH GARDEN 1999-10-13    None   
147845            SALAK SOUTH GARDEN 1999-09-13    None   
147850            SALAK SOUTH GARDEN 1999-07-19    None   
155027             TAMAN DESA CHERAS 1994-06-07    None   
165318               BAIDURI BLOCK G 2001-01-08    None   
165380               BAIDURI BLOCK G 1997-09-16    None   
186764          10 MONT KIARA @ MK10 2010-05-26    None   
203050            MONT KIARA MERIDIN 2010-07-28    None   
206358                    KIARA 1888 2010-05-13    None   
218018                      SUTRAMAS 2010-03-31    None   
224717                 BAYVIEW COURT 1998-05-26    None   

                       building_type     tenure floors rooms  land_area  \
18211   TERRACE HOUSE - INTERMEDIATE  LEASEHOLD      2   3.0  1,647 ft²   
27824   TERRACE HOUSE - INTERMEDIATE   FREEHOLD      2   3.0  1,650 ft²   
70435                       BUNGALOW  LEASEHOLD      1   3.0  4,004 ft²   
71295   TERRACE HOUSE - INTERMEDIATE   FREEHOLD      1   3.0  1,754 ft²   
71296   TERRACE HOUSE - INTERMEDIATE   FREEHOLD      1   3.0  1,775 ft²   
74962   TERRACE HOUSE - INTERMEDIATE   FREEHOLD      2   3.0  1,787 ft²   
75638                         SEMI-D  LEASEHOLD      2   4.0  3,057 ft²   
102998                          FLAT   FREEHOLD      1   3.0    708 ft²   
112720                          FLAT   FREEHOLD      1   nan    734 ft²   
147816                          FLAT   FREEHOLD      1     2    625 ft²   
147843                          FLAT   FREEHOLD      1     2    607 ft²   
147845                          FLAT   FREEHOLD      1     2    610 ft²   
147850                          FLAT   FREEHOLD      1     2    611 ft²   
155027                          FLAT   FREEHOLD      1     2    625 ft²   
165318                     APARTMENT  LEASEHOLD      1     3    945 ft²   
165380                     APARTMENT  LEASEHOLD      1     3    878 ft²   
186764                   CONDOMINIUM   FREEHOLD      1     5   4091 ft²   
203050                   CONDOMINIUM   FREEHOLD      1     3   2512 ft²   
206358                   CONDOMINIUM   FREEHOLD      1     3   1238 ft²   
218018                   CONDOMINIUM   FREEHOLD      1     4   1377 ft²   
224717                     APARTMENT  LEASEHOLD      1     3    750 ft²   

         built_up price_psf      price  
18211   1,087 ft²       107    175,750  
27824   1,671 ft²       262    432,000  
70435     834 ft²        30    120,000  
71295     869 ft²       114    200,000  
71296     869 ft²       113    200,000  
74962   1,323 ft²       123    220,000  
75638   2,485 ft²       212    648,000  
102998    708 ft²       132     93,500  
112720    734 ft²       109     80,000  
147816    625 ft²        56     35,000  
147843    607 ft²        92     55,800  
147845    610 ft²       109     66,500  
147850    611 ft²        95     58,000  
155027    625 ft²        48     30,000  
165318    945 ft²       146    138,000  
165380    878 ft²       148    130,000  
186764   4091 ft²       660  2,700,000  
203050   2512 ft²       430  1,080,000  
206358   1238 ft²       564    698,000  
218018   1377 ft²       305    420,000  
224717    750 ft²        80     60,000

We can drop the `address` column based on the following reasons:
1. No detailed locational scope is considered in this study
2. Only high-level geographical scope is considered

In [7]:
df_transactions['building_type'].value_counts()

building_type
CONDOMINIUM                      99778
TERRACE HOUSE - INTERMEDIATE     47865
APARTMENT                        45264
FLAT                             44960
SERVICE RESIDENCE                27787
BUNGALOW                          7871
SEMI-D                            5824
TOWN HOUSE                        4983
TERRACE HOUSE - CORNER LOT        3946
TERRACE HOUSE - END LOT           2966
CLUSTER HOUSE - INTERMEDIATE      2588
CLUSTER HOUSE                      506
CLUSTER HOUSE - CORNER LOT         128
CLUSTER HOUSE - END LOT             99
CONDOMINIUM - CLUB                   1
CONDOMINIUM - RESIDENTIAL USE        1
Name: count, dtype: int64

Findings:
1. Majority of the transactions are condominium, followed by intermediate terrace house and apartment
2. Cluster house has an unknown category which does not fall into intermediate, corner lot or end lot.
3. Condominium has a weird category called "residential use", but only has one transaction

In [8]:
print(f"% Freehold: {df_transactions['tenure'].value_counts()['FREEHOLD'] / len(df_transactions) * 100}")
df_transactions['tenure'].value_counts()

% Freehold: 61.22987300003055


tenure
FREEHOLD     180363
LEASEHOLD    114204
Name: count, dtype: int64

Findings:
1. Majority of the transactions are freehold (61.23%)

In [9]:
df_transactions['floors'].value_counts()

floors
1     232497
2      44132
3      11788
2½      3059
0       1175
4        967
3½       264
1½       256
6        154
5        150
8         41
4½        18
9         17
20        15
10        11
7          7
11         4
13         2
15         2
99         2
24         2
46         1
12         1
35         1
21         1
Name: count, dtype: int64

In [10]:
print(df_transactions['floors'].isnull().sum())
print((df_transactions['floors'] == 'nan').sum())

0
0


Findings:
1. Half a floor are represented by 1/2 instead of .5
2. There are weird values like:
    - 0 floor
    - 99 floors
3. Majority of floors are 2 floors
4. No missing values

In [11]:
df_transactions['rooms'].value_counts()

rooms
3       104800
3.0      56962
2        30944
nan      30670
4.0      20333
         ...  
31           1
17.0         1
24.0         1
18.0         1
148          1
Name: count, Length: 70, dtype: int64

In [12]:
(df_transactions['rooms'] == 'nan').sum() / len(df_transactions) * 100

10.411892710317177

Findings:
1. Weird number of rooms:
    - 0 room
    - 46 rooms
2. Majority of the transactions are 3 rooms
3. 10.45% of the transactions have NaN

In [13]:
df_transactions['land_area'].value_counts()

land_area
1,647 ft²     3652
883 ft²       3218
1,539 ft²     2388
592 ft²       2146
1,540 ft²     1882
              ... 
10,807 ft²       1
2,202 ft²        1
25,306 ft²       1
4,690 ft²        1
5593 ft²         1
Name: count, Length: 9147, dtype: int64

In [14]:
df_transactions['land_area'].isnull().sum() / len(df_transactions) * 100

0.0

Findings:
1. There are "ft2" and "," in land_area. Need to remove.
2. No missing values

In [15]:
df_transactions['built_up'].value_counts()

built_up
nan          5613
861 ft²      1849
1000 ft²     1670
549 ft²      1452
556 ft²      1445
             ... 
1,031 ft²       1
113 ft²         1
5,414 ft²       1
5197 ft²        1
5593 ft²        1
Name: count, Length: 7610, dtype: int64

In [16]:
(df_transactions['built_up'] == 'nan').sum() / len(df_transactions) * 100

1.9055087637108028

Findings:
1. 1.91% of the data is missing. Need further investigation
2. There are "ft2" and "," in the data. Need to remove them

In [17]:
df_transactions['price_psf'].value_counts()

price_psf
152      1246
155      1241
71       1123
139      1104
143      1092
         ... 
1,645       1
1,980       1
3,099       1
1,578       1
2,326       1
Name: count, Length: 2144, dtype: int64

In [18]:
df_transactions['price_psf'].isna().sum() / len(df_transactions) * 100

0.0

Findings:
1. No missing values
2. Have "," in the numbers. Need to remove them

In [19]:
df_transactions['price'].value_counts()

price
200,000    3603
150,000    3587
300,000    3511
250,000    3382
500,000    3181
           ... 
101,800       1
123,620       1
73,360        1
554,338       1
518,296       1
Name: count, Length: 18778, dtype: int64

In [20]:
df_transactions['price'].isna().sum() / len(df_transactions) * 100

0.0

Findings:
1. No missing data.
2. Have "," in the data. Need to remove.

### Concluding Remarks
1. How does each dataset look like? Is it in row format or it has merged cells?
    - It is in row format, with no merged cells.
2. What is the start and end of each dataset?
    - From 1909 to 2023
3. Is the dataset from the area of interest, Kuala Lumpur?
    - Yes
4. How is the data quality?
    - There are missing values in some columns (`built_up` and `rooms`). Need to decide to remove rows or impute.
    - Values in some columns (`land_area` and `built_up`) has unit of measurement like "ft2". Need to remove.
    - Values in `floors` column uses fraction instead of decimal. Need to change.
    - Values in numerical columns (`land_area`, `built_up`, `price_psf` and `price`) has comma. Need to remove.
    - Some values in some columns (`floors` and `rooms`) does not make sense. Need further investigation.
    - Only landed properties are in the dataset. Need another round of scraping for high rise real estate.

In [21]:
df_transactions.to_parquet(RAW_DATA_DIR / 'transactions_KL_ckpt1.parquet', index=False)